### ISSUE TITLE
Real time UI editing with Firefox debugging protocol

### ISSUE THREAD

**Issue by bovine3dom** [tridactyl/tridactyl#460]
This is what the "browser toolbox" uses to provide the ability to edit `userChrome.css` without a restart.

There's no reason that we couldn't implement this with the native messenger, but it looks like it would be at least a little complicated. The easiest thing to do would be to listen to what the debugger sends when the userChrome is edited to make the browser reload it.

https://searchfox.org/mozilla-central/source/devtools/docs/backend/protocol.md

**Comment by glacambre** [comment]
Since their tool is all JS we could also grab their implementation of the protocol, have Tridactyl connect to the native messenger through a websocket and have the native messenger forward Tridactyl's packets to Firefox.

**Comment by glacambre** [comment]
Apparently we wouldn't even need to go through the native messenger, Firefox accepts websocket connections: https://searchfox.org/mozilla-central/source/devtools/client/framework/target-from-url.js#125

**Comment by bovine3dom** [comment]
Huh, so we could potentially do a bunch of the native stuff with it, e.g navigating to about:urls? Maybe even intercepting key events on restricted pages? The downside being the scary "a program is trying to control Firefox" warning every session.

**Comment by glacambre** [comment]
Setting `devtools.debugger.prompt-connection` to false disables this warning. We might be able to do really interesting things with this!

**Comment by jjeffrey** [comment]
Has there been any progress toward implementing this real-time UI editing?

**Comment by glacambre** [comment]
No, there hasn't. You can do live editing by using the browser toolbox though (top-right hamburger menu -> more tools -> browser toolbox).
